File: backend/utils/validation.py
"""Input validation for prediction requests."""

from typing import Any


class ValidationError(Exception):
    """Raised when input validation fails."""

    def __init__(self, errors: list[dict[str, str]]):
        self.errors = errors
        super().__init__(f"Validation failed: {errors}")


VALID_CATEGORIES = {
    'gender': ['Male', 'Female'],
    'marital_status': ['Married', 'Single', 'Divorced', 'Widowed'],
    'family_status': ['Civil Marriage', 'Married', 'Single / Not Married', 'Separated', 'Widow'],
    'education': ['Higher Education', 'Secondary', 'Incomplete Higher', 'Lower Secondary', 'Academic Degree'],
    'income_type': ['Working', 'Commercial Associate', 'Pensioner', 'State Servant', 'Student'],
    'employment_status': ['Employed', 'Self-employed', 'Unemployed', 'Maternity Leave'],
    'housing_type': ['House / Apartment', 'With Parents', 'Municipal Apartment', 'Rented Apartment', 'Office Apartment', 'Co-op Apartment'],
    'loan_payment_status': ['On Time', 'Delayed', 'Critical'],
}

REQUIRED_FIELDS = [
    'gender', 'age', 'marital_status', 'family_status', 'children',
    'education', 'income_type', 'employment_status', 'employment_duration_years',
    'annual_income', 'housing_type', 'owns_car', 'owns_realty',
    'existing_loans', 'existing_debt', 'loan_payment_status',
    'credit_inquiries', 'credit_history_years', 'payment_history_score',
]

NUMERIC_RANGES = {
    'age': (18, 100),
    'employment_duration_years': (0, 60),
    'annual_income': (0, 10_000_000),
    'children': (0, 20),
    'existing_loans': (0, 50),
    'existing_debt': (0, 5_000_000),
    'credit_inquiries': (0, 30),
    'credit_history_years': (0, 60),
    'payment_history_score': (0, 1),
}


def validate_applicant(data: dict[str, Any]) -> list[dict[str, str]]:
    """Validate applicant input. Returns a list of error dicts."""
    errors = []

    # Check required fields
    for field in REQUIRED_FIELDS:
        if field not in data or data[field] is None:
            errors.append({'field': field, 'message': f'{field} is required.'})

    if errors:
        return errors

    # Check numeric ranges
    for field, (min_val, max_val) in NUMERIC_RANGES.items():
        value = data.get(field)
        if value is None:
            continue
        try:
            v = float(value)
            if v < min_val or v > max_val:
                errors.append({
                    'field': field,
                    'message': f'{field} must be between {min_val} and {max_val}.',
                })
        except (TypeError, ValueError):
            errors.append({'field': field, 'message': f'{field} must be a number.'})

    # Check categorical values
    for field, valid_values in VALID_CATEGORIES.items():
        value = data.get(field)
        if value is not None and value not in valid_values:
            errors.append({
                'field': field,
                'message': f'{field} must be one of: {", ".join(valid_values)}.',
            })

    # Check boolean fields
    for field in ['owns_car', 'owns_realty']:
        value = data.get(field)
        if value is not None and not isinstance(value, bool):
            errors.append({'field': field, 'message': f'{field} must be true or false.'})

    return errors

File: src/lib/validation.ts
import type { ApplicantInput } from '@/types/applicant';

export interface ValidationError {
  field: string;
  message: string;
}

export function validateApplicant(input: Partial<ApplicantInput>): ValidationError[] {
  const errors: ValidationError[] = [];

  if (input.age === undefined || input.age < 18 || input.age > 100) {
    errors.push({ field: 'age', message: 'Age must be between 18 and 100.' });
  }
  if (input.annualIncome === undefined || input.annualIncome < 0 || input.annualIncome > 10_000_000) {
    errors.push({ field: 'annualIncome', message: 'Annual income must be between $0 and $10,000,000.' });
  }
  if (input.employmentDurationYears === undefined || input.employmentDurationYears < 0 || input.employmentDurationYears > 60) {
    errors.push({ field: 'employmentDurationYears', message: 'Employment duration must be between 0 and 60 years.' });
  }
  if (input.children === undefined || input.children < 0 || input.children > 20) {
    errors.push({ field: 'children', message: 'Number of children must be between 0 and 20.' });
  }
  if (input.existingLoans !== undefined && (input.existingLoans < 0 || input.existingLoans > 50)) {
    errors.push({ field: 'existingLoans', message: 'Existing loans must be between 0 and 50.' });
  }
  if (input.existingDebt !== undefined && (input.existingDebt < 0 || input.existingDebt > 5_000_000)) {
    errors.push({ field: 'existingDebt', message: 'Existing debt must be between $0 and $5,000,000.' });
  }
  if (input.creditInquiries !== undefined && (input.creditInquiries < 0 || input.creditInquiries > 30)) {
    errors.push({ field: 'creditInquiries', message: 'Credit inquiries must be between 0 and 30.' });
  }
  if (input.creditHistoryYears !== undefined && (input.creditHistoryYears < 0 || input.creditHistoryYears > 60)) {
    errors.push({ field: 'creditHistoryYears', message: 'Credit history must be between 0 and 60 years.' });
  }
  if (input.paymentHistoryScore !== undefined && (input.paymentHistoryScore < 0 || input.paymentHistoryScore > 1)) {
    errors.push({ field: 'paymentHistoryScore', message: 'Payment history score must be between 0 and 1.' });
  }

  return errors;
}
